bcm53xx: add support for ASUS RT-AC3200 and RT-AC5300

tbrautaset opened the pull request:
## [Overview]

Adds initial OpenWrt support for the **ASUS RT-AC3200** and **ASUS RT-AC5300**.

Both are **BCM4709-based** devices (already supported under `bcm53xx`), with NAND flash and an integrated Gigabit switch.

- [RT-AC3200 hardware page](https://openwrt.org/toh/hwdata/asus/asus_rt-ac3200)  
- [RT-AC5300 hardware page](https://openwrt.org/toh/hwdata/asus/asus_rt-ac5300)

**Status**  

[OK] Bootable via CFE MiniWeb recovery and `sysupgrade`  

[OK] Ethernet LAN/WAN verified working  

[OK] Wi-Fi consistent with other BCM4709-based bcm53xx devices  

[NOTE] On **RT-AC3200**, radios only appear after reboot with the bcm43602_variables quirk applied.  
Without this one-time init, no radios are detected on first boot.  

Inside ASUS OEM firmware for ASUS RT-AC5300, the standard `dhd.ko` only includes support for **BCM4366C0**.  
Code paths for both **BCM4366B1** and **BCM4366C0** exist in `dhd24.ko` with much older firmware revisions.

---

## [Changes]

- Added DTS files: `bcm4709-asus-rt-ac3200.dts`, `bcm47094-asus-rt-ac5300.dts`
- Added image definitions under `target/linux/bcm53xx/image/Makefile`
- Added dedicated NVRAM handling:
  - Extended `nvram-bcm53xx.init` for proper Wi-Fi LED behavior
  - No USB Wi-Fi hardware present → ideally `brcmfmac-usb` should also be dropped; for now handled in `.config` with:  
    ```make
    # CONFIG_BRCMFMAC_USB is not set
    ```

**Files changed (6)**  
1. `package/utils/nvram/Makefile` — bump release  
3. `package/utils/nvram/files/nvram-bcm53xx.init` — extend LED handling for AC3200/AC5300 Wi-Fi LEDs  
4. `target/linux/bcm53xx/patches-6.6/311-ARM-BCM5301X-Add-DT-for-Asus-RT-AC3200.patch` — DTS for AC3200  
5. `target/linux/bcm53xx/patches-6.6/312-ARM-BCM5301X-Add-DT-for-Asus-RT-AC5300.patch` — DTS for AC5300  

---

## [Hardware Specs]

|                        | **ASUS RT-AC3200**                         | **ASUS RT-AC5300**                          |
|------------------------|--------------------------------------------|---------------------------------------------|
| SoC                    | Broadcom **BCM4709A0**, 2c @ **1.0 GHz**   | Broadcom **BCM4709C0**, 2c @ **1.4 GHz**    |
| RAM / Flash            | **256 MB DDR3** / **128 MB NAND**          | **512 MB DDR3** / **128 MB NAND**           |
| Radios                 | **3 × BCM43602** (1×2.4 G, 2×5 G)          | **3 × BCM4366C0** (1×2.4 G, 2×5 G)          |
| Ethernet               | **1× WAN + 4× LAN** (Gigabit)              | **1× WAN + 4× LAN** (Gigabit)               |
| USB                    | **USB2 ×1**, **USB3 ×1**                   | **USB2 ×1**, **USB3 ×1**                    |
| USB Wi-Fi / LEDs       | **Not present**                            | **Not present**                             |

---

## [Optional A/B Testing]

This section is outside the scope of this PR.  
Firmware updates are handled in `package/firmware/linux-firmware`.

Repo: [asus-brcm43602-brcm4366c](https://github.com/tbrautaset/asus-brcm43602-brcm4366c), Files:
  - brcmfmac43602-pcie.bin — 7.35.177.61 (from linux-firmware; newer than current OpenWrt)  
  - brcmfmac4366c-pcie.bin — 10.10.122.303 (OEM test blob; not suitable for inclusion)  

---

# Build-time override (on your build host)

mkdir -p files/lib/firmware/brcm

curl -fL -o files/lib/firmware/brcm/brcmfmac43602-pcie.bin&#x20;
[https://raw.githubusercontent.com/tbrautaset/asus-brcm43602-brcm4366c/main/brcmfmac43602-pcie.bin](https://raw.githubusercontent.com/tbrautaset/asus-brcm43602-brcm4366c/main/brcmfmac43602-pcie.bin)

curl -fL -o files/lib/firmware/brcm/brcmfmac4366c-pcie.bin&#x20;
[https://raw.githubusercontent.com/tbrautaset/asus-brcm43602-brcm4366c/main/brcmfmac4366c-pcie.bin](https://raw.githubusercontent.com/tbrautaset/asus-brcm43602-brcm4366c/main/brcmfmac4366c-pcie.bin)

sha256sum files/lib/firmware/brcm/brcmfmac436\*.bin

---

# Runtime override (copy to a running router)

ssh root\@router 'mkdir -p /lib/firmware/brcm'

scp brcmfmac43602-pcie.bin  root\@router:/lib/firmware/brcm/
scp brcmfmac4366c-pcie.bin  root\@router:/lib/firmware/brcm/
ssh root\@router 'sha256sum /lib/firmware/brcm/brcmfmac436\*.bin; sync; reboot'

---

Notes:

* `brcmfmac43602-pcie.bin` v7.35.177.61 is upstream in linux-firmware, but OpenWrt
  currently ships an older revision.
* The 4366C OEM blob cannot be included in OpenWrt due to licensing and is only
  provided here for voluntary A/B testing.

---

## [Acknowledgements]

* Replaces PR #19373
* Co-developed with @arinc9

---

Thanks for reviewing.

przemekp1 commented:
Hi, is it possible to expect support for the RT-AC2900 model?
ETH chip1: Broadcom BCM4906
Switch: Broadcom BCM4906

tbrautaset commented:
> Hi, is it possible to expect support for the RT-AC2900 model? ETH chip1: Broadcom BCM4906 Switch: Broadcom BCM4906

Hi,

The RT-AC2900 is based on the BCM4906 platform, which is not the same as the BCM4709 used in the AC3200/AC5300.

Specifically:

-     BCM4906 belongs to the newer Broadcom BCM4908/4906/6750 family (Northstar 2).

-     It uses a 64-bit ARM Cortex-A53 CPU (arm64).

-     It is targeted by the bcm4908 platform in OpenWrt (not bcm53xx).

At the moment, mainline OpenWrt does not have stable support for BCM4906/BCM4908, though there has been experimental work and some progress in forks (e.g., Asuswrt-Merlin, early OpenWrt branches).

If this model were supported today, it would show up here:
https://downloads.openwrt.org/releases/24.10.2/targets/bcm4908/generic/

Since it’s not listed there currently, you (or someone who owns this device) would need to:

-     Port it by preparing a device tree and platform definitions.

-     Test and verify flash layout and bootloader.

-     Submit a Pull Request to get it reviewed and merged into OpenWrt.

So unfortunately, support for the RT-AC2900 isn’t included in this PR or in bcm53xx.

Hope this clarifies!

tbrautaset commented:
No rush – just noting the “Test Formalities” check has been queued for several hours.

seuros commented on target/linux/bcm53xx/patches-6.6/311-ARM-BCM5301X-Add-DT-for-Asus-RT-AC3200.patch:
```diff
@@ -0,0 +1,170 @@
++			linux,default-trigger = "default-on";
++		};
++
++		led-wan-red {
```
What i mean is that unless you have a limited series model or that i'm colorblind every led is white: 
https://www.blacktubi.com/review/asus-rt-ac5300-review/

If we have to mark it red due to openwrt usage , that is fine.

From my tests, we have control to 4/6 leds.

arinc9 replied:
@seuros I believe a red WAN LED exists along with a white one on this device. So the devicetree is correct. Also, the devicetree patch is in Linux mainline so making changes in the devicetree patch is out of scope of this PR, in any case.

tbrautaset replied:
@seuros According to the **SmallNetBuilder schematic**:
🔗 [https://www.smallnetbuilder.com/wp-content/uploads/2016/01/asus\_rtac5300\_callouts.jpg](https://www.smallnetbuilder.com/wp-content/uploads/2016/01/asus_rtac5300_callouts.jpg)

The **WAN (Internet) LED** behaves as follows:

* **Red** → No IP or no physical connection
* **On** (white/green) → Valid physical WAN connection

So the all-white LEDs you see are actually expected in normal operation — no “limited series” model and no colorblindness required 😄

tbrautaset committed:
bcm53xx: add device tree for ASUS RT-AC3200

Add the device tree for ASUS RT-AC3200 which is an AC3200 router featuring
5 Ethernet ports over the integrated Broadcom switch.

Hardware info:
* Processor: Broadcom BCM4709A0 dual-core @ 1.0 GHz
* Switch: BCM53012 in BCM4709A0
* DDR3 RAM: 256 MB
* Flash: 128 MB
* 2.4GHz: BCM43602 3x3 single chip 802.11b/g/n SoC
* 5GHz: BCM43602 3x3 two chips 802.11a/n/ac SoC
* Ports: 4 LAN Ports, 1 WAN Port

Co-developed-by: Tom Brautaset <tbrautaset@gmail.com>
Signed-off-by: Tom Brautaset <tbrautaset@gmail.com>
Signed-off-by: Arınç ÜNAL <arinc.unal@arinc9.com>

tbrautaset committed:
bcm53xx: add device tree for ASUS RT-AC5300

Add the device tree for ASUS RT-AC5300 which is an AC5300 router featuring
5 Ethernet ports over the integrated Broadcom switch.

Hardware info:
* Processor: Broadcom BCM4709C0 dual-core @ 1.4 GHz
* Switch: BCM53012 in BCM4709C0
* DDR3 RAM: 512 MB
* Flash: 128 MB
* 2.4GHz: BCM4366 4x4 single chip 802.11b/g/n SoC
* 5GHz: BCM4366 4x4 two chips 802.11a/n/ac SoC
* Ports: 4 LAN Ports, 1 WAN Port

Co-developed-by: Tom Brautaset <tbrautaset@gmail.com>
Signed-off-by: Tom Brautaset <tbrautaset@gmail.com>
Signed-off-by: Arınç ÜNAL <arinc.unal@arinc9.com>

tbrautaset committed:
package/nvram: add ASUS RT-AC3200 defaults and firstboot init

Hello.

This patch series brings support for ASUS RT-AC3200 and ASUS RT-AC5300 on
OpenWrt.

Added a dedicated NVRAM package for ASUS RT-AC3200 (aligned with RT-AC5300)
so that all three radios are detected.

Introduced a one-time 04-nvram-firstboot script: applies NVRAM.

Note to Rafał, please apply this after backporting the device tree patches
for these devices.

Co-developed-by: Tom Brautaset <tbrautaset@gmail.com>
Signed-off-by: Tom Brautaset <tbrautaset@gmail.com>
Signed-off-by: Arınç ÜNAL <arinc.unal@arinc9.com>

tbrautaset committed:
bcm53xx: refine device packages for ASUS RT-AC3200/RT-AC5300

- Select correct brcmfmac firmware per radio/stepping:
  * RT-AC3200 → BCM43602A1 (brcmfmac43602-pcie.bin)
  * RT-AC5300 → BCM4366C0 (brcmfmac4366c-pcie.bin)
    (dhd24.ko bundles both B1/C0, but AC5300 devices request C0; if a
     device ever requests brcmfmac4366b-pcie.bin in dmesg, we can add B1)

- Drop kmod-usb-ledtrig-usbport (not used on these targets).

- Note: brcmfmac-usb is not needed (no USB Wi-Fi on these devices). It is
  currently kept disabled via .config (# CONFIG_BRCMFMAC_USB is not set);
  follow-up can remove it from device packages explicitly.

Co-developed-by: Tom Brautaset <tbrautaset@gmail.com>
Signed-off-by: Tom Brautaset <tbrautaset@gmail.com>
Signed-off-by: Arınç ÜNAL <arinc.unal@arinc9.com>
